fakenet/listeners/ProxyListener.py

```diff
@@ -0,0 +1,397 @@
+
+        self.server.logger.debug('Received UDP packet from %s.' % 
+                self.client_address[0])
+
+        if data:
+
+            self.server.logger.debug('Packet data\n%s' % hexdump.hexdump(data, 
+                result='return'))
+
+            orig_src_ip = self.client_address[0]
+            orig_src_port = self.client_address[1]
+
+            top_listener = get_top_listener(self.server.config, data, 
+                    self.server.listeners, self.server.diverter, 
+                    orig_src_ip, orig_src_port, 'UDP')
+
+            if top_listener:
+
+                # queue for data received from the listener
+                listener_q = Queue.Queue()
+                # queue for data received from remote
+                remote_q = Queue.Queue()
+                self.server.logger.debug('Likely listener: %s' % 
+                        top_listener.name)
+                listener_sock = ThreadedUDPClientSocket('localhost', 0, 
+                        listener_q, remote_q, self.server.config, 
+                        self.server.logger, top_listener.port, 
+                        self.server.fwd_table, orig_src_ip, orig_src_port)
+                listener_sock.setDaemon(True)
+                listener_sock.start()
+                remote_sock.setblocking(0)
+
+                # no peek option so process the data already recd
+                remote_q.put(data)
+
+                try:
+                    while True:
+                        readable, writable, exceptional = select.select(
+                                [remote_sock], [], [], .001)
+                        if readable:
+                            data = remote_sock.recv(BUF_SZ)
+                            if data:
+                                remote_q.put(data)
+                            else:
+                                self.server.logger.debug(
+                                        'Closing remote socket connection')
+                                return
+                        if not listener_q.empty():
+                            data = listener_q.get()
+                            remote_sock.sendto(data, (orig_src_ip, int(orig_src_port)))
+                except Exception as e:
+                    pass
+
+        else:
+            self.server.logger.debug('No packet data')
+
+def main():
+
+    logging.basicConfig(format='%(asctime)s [%(name)15s] %(message)s', 
+            datefmt='%m/%d/%y %I:%M:%S %p', level=logging.DEBUG)
+    global listeners
+    listeners = load_plugins()
+
+    TCP_server = ThreadedTCPServer((IP, int(sys.argv[1])), 
+            ThreadedTCPRequestHandler)
+    TCP_server_thread = threading.Thread(target=TCP_server.serve_forever)
+    TCP_server_thread.daemon = True
+    TCP_server_thread.start()
+    tcp_server_ip, tcp_server_port = TCP_server.server_address
+    logger.info("TCP Server(%s:%d) thread: %s" % (tcp_server_ip, 
+        tcp_server_port, TCP_server_thread.name))
+
+    try:
```

Comment:
I know this is just test code but if you wanted to remove/simplify some code, you can remove the thread here. When `serve_forever` accepts a new connection/request, it'll spawn a new thread to handle it. It'll block and periodically poll to see if needs to shutdown.

Reply:
I think the testing code needs to be removed entirely. The proxy makes no sense if it doesn't know about the other listeners, which only exist if it is run through fakenet.py